AbigailDawson/knownative
# KN-9: Add ESLint and Prettier

### Describe your changes
Added ESLint and Prettier to project to automatically format and look for common mistakes. Runs automatically when working in files but can be triggered in the terminal to cover entire project. For prettier, run `npm run format` and for ESLint errors, run `npm run lint.` 

Because this PR adds new dependencies, you will probably need to run `npm i` to install all necessary dependencies.

Prettier: 
You can edit the rules implemented by Prettier in the `.prettierrc` file. For all available options, go to [Prettier Docs](https://prettier.io/docs/en/options)
There is also a `.prettierignore` file which limits prettiers reach. It will not format anything labeled in the prettierignore file.

Devs may need to go into their workspace settings and turn on 'Format on Save' for the auto formatting to work. 
Also, the Prettier and the ESLint extensions may be helpful to install on your code platform, but I don't believe are required.

ESLint: 
You can edit the rules implemented by ESLint in the `eslint.config.mjs` file in the rules objects. Ive added 3 I find useful on top of the basic checks ESLint does. For all available rules, go to [ESLint Docs](https://eslint.org/docs/latest/rules/). You can also limit what files eslint can check in this file in the "ignores" list. Something to consider is to install a config from a well established company (airbnb, facebook, shopify... heres a [good list](https://github.com/dustinspecker/awesome-eslint?tab=readme-ov-file#configs-by-well-known-companiesorganizations)) but these may be more strict than the current bare bones this PR implements..

### Issue ticket number and link:
https://huly.app/workbench/knownative/tracker/KN-9

### Notes
This PR does not reformat the entire project using prettier. A second PR will be opened after the next release when there are fewer devs working to avoid conflicts. 

### How to test
Test Prettier in a File:
> Open a Jsx file. 
> Code using random spaces, double quotes, no semis. 
> Save file. 
> File should automatically format itself on save. If it doesn't, Devs may need to go into their workspace settings and turn on 'Format on Save' for the auto formatting to work. 
> ![Screenshot 2024-09-11 at 11 35 30 AM](https://github.com/user-attachments/assets/af151ec5-1925-4247-9bcc-9e05a51f3172)

Test Prettier in Terminal (NOTE: Formats the entire project):
> Open a new tab in Terminal  (make sure youre in the knownative directory)
> Run `npm run format`
> Terminal will print all files it formatted. You will also notice the changes in git
> <img width="1312" alt="Screenshot 2024-09-11 at 11 57 45 AM" src="https://github.com/user-attachments/assets/855537ce-45b1-4dd3-aeb6-beecef9964b3">

Test ESLint in file:
> Open a JSX file
> Add a console log.
> Notice that the console log displays an ERROR line under the console log
> Go to a file like DemoInfoSidebar.jsx and notice how unused variables are marked as errors, as well as `class`, which should be changed to `className`
> ![Screenshot 2024-09-11 at 12 05 13 PM](https://github.com/user-attachments/assets/49574999-ede1-476a-83c0-69bb78c67ec4)

Test ESLint in Terminal:
> Open a new tab in Terminal  (make sure youre in the knownative directory)
> Run `npm run lint`
> Terminal will print all errors it finds that need to be corrected and the file they are in
> Note: The terminal that also deploys the frontend will also display errors that should be addressed
> <img width="401" alt="Screenshot 2024-09-11 at 12 08 22 PM" src="https://github.com/user-attachments/assets/7b85be92-3ad4-44e2-9b64-21919927b932">
